### PR TITLE
[PDI-19926] Put a file with FTP: An error doesn't occur when we change the local directory permissions.

### PR DESCRIPTION
**Pull request by Mandakini111**

### CODE REVIEW

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 8m 22s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl \
plugins/put-a-file-with-ftp/impl
```
---
:ok_hand: All tests passed!
 
Tests run: **5**, Failures: **0**, Skipped: **0** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/pdi/job/PR-9078/1/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

File: plugins/put-a-file-with-ftp/impl/src/main/java/org/pentaho/di/job/entries/ftpput/JobEntryFTPPUT.java [hunk 1]
```diff
@@ -639,8 +642,16 @@ public Result execute( Result previousResult, int nr ) {
         }
       }
 
+      Path path = Paths.get( realLocalDirectory );
```
Comment (NJtwentyone) [inline plugins/put-a-file-with-ftp/impl/src/main/java/org/pentaho/di/job/entries/ftpput/JobEntryFTPPUT.java:645]: You need to add unit tests for this new edge cases for the logic added in JobEntryFTPPUT.java#execute
There is an existing class:
https://github.com/pentaho/pentaho-kettle/blob/master/plugins/put-a-file-with-ftp/impl/src/test/java/org/pentaho/di/job/entries/ftpput/JobEntryFTPPUTTest.java

I understand the function is 200+ lines with many many edge cases. I'm only asking the two exceptions you just added.

**Review by NJtwentyone — CHANGES_REQUESTED** [review]
Add unit tests

**Comment by buildguy** [comment]
### :warning: Build finished in 6m 14s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl \
plugins/put-a-file-with-ftp/impl
```
---
### :no_entry: Failed Tests

#### :cloud_with_lightning_and_rain: 3 test(s) failed:

<details><summary><b>org.pentaho.di.job.entries.ftpput.JobEntryFTPPUTLoadSaveTest.testSerialization</b> (click to expand)</summary><p>

```
Error validating password
```

</p></details>

<details><summary><b>org.pentaho.di.job.entries.ftpput.JobEntryFTPPUTTest.initialSetUp</b> (click to expand)</summary><p>

```

For org.pentaho.di.core.encryption.Encr, static mocking is already registered in the current thread

To create a new mock, the existing static mock registration must be deregistered
```

</p></details>

<details><summary><b>org.pentaho.di.job.entries.ftpput.JobEntryFTPPUTTest.testNoFilePermissions</b> (click to expand)</summary><p>

```

For org.pentaho.di.core.encryption.Encr, static mocking is already registered in the current thread

To create a new mock, the existing static mock registration must be deregistered
```

</p></details>

  
Tests run: **8**, Failures: **3**, Skipped: **0** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/pdi/job/PR-9078/2/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :warning: Build finished in 3m 11s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl \
plugins/put-a-file-with-ftp/impl
```
---
### :no_entry: Failed Tests

#### :cloud_with_lightning_and_rain: 3 test(s) failed:

<details><summary><b>org.pentaho.di.job.entries.ftpput.JobEntryFTPPUTLoadSaveTest.testSerialization</b> (click to expand)</summary><p>

```
Error validating password
```

</p></details>

<details><summary><b>org.pentaho.di.job.entries.ftpput.JobEntryFTPPUTTest.initialSetUp</b> (click to expand)</summary><p>

```

For org.pentaho.di.core.encryption.Encr, static mocking is already registered in the current thread

To create a new mock, the existing static mock registration must be deregistered
```

</p></details>

<details><summary><b>org.pentaho.di.job.entries.ftpput.JobEntryFTPPUTTest.testNoFilePermissions</b> (click to expand)</summary><p>

```

For org.pentaho.di.core.encryption.Encr, static mocking is already registered in the current thread

To create a new mock, the existing static mock registration must be deregistered
```

</p></details>

  
Tests run: **8**, Failures: **3**, Skipped: **0** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/pdi/job/PR-9078/3/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :warning: Build finished in 3m 22s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl \
plugins/put-a-file-with-ftp/impl
```
---
### :no_entry: Failed Tests

#### :cloud_with_lightning_and_rain: 3 test(s) failed:

<details><summary><b>org.pentaho.di.job.entries.ftpput.JobEntryFTPPUTLoadSaveTest.testSerialization</b> (click to expand)</summary><p>

```
Error validating password
```

</p></details>

<details><summary><b>org.pentaho.di.job.entries.ftpput.JobEntryFTPPUTTest.initialSetUp</b> (click to expand)</summary><p>

```

For org.pentaho.di.core.encryption.Encr, static mocking is already registered in the current thread

To create a new mock, the existing static mock registration must be deregistered
```

</p></details>

<details><summary><b>org.pentaho.di.job.entries.ftpput.JobEntryFTPPUTTest.testNoFilePermissions</b> (click to expand)</summary><p>

```

For org.pentaho.di.core.encryption.Encr, static mocking is already registered in the current thread

To create a new mock, the existing static mock registration must be deregistered
```

</p></details>

  
Tests run: **8**, Failures: **3**, Skipped: **0** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/pdi/job/PR-9078/4/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :warning: Build finished in 3m 28s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl \
plugins/put-a-file-with-ftp/impl
```
---
### :no_entry: Failed Tests

#### :cloud_with_lightning_and_rain: 2 test(s) failed:

<details><summary><b>org.pentaho.di.job.entries.ftpput.JobEntryFTPPUTLoadSaveTest.testSerialization</b> (click to expand)</summary><p>

```
Error validating password
```

</p></details>

<details><summary><b>org.pentaho.di.job.entries.ftpput.JobEntryFTPPUTTest.testNoFilePermissions</b> (click to expand)</summary><p>

```

For org.pentaho.di.core.encryption.Encr, static mocking is already registered in the current thread

To create a new mock, the existing static mock registration must be deregistered
```

</p></details>

  
Tests run: **7**, Failures: **2**, Skipped: **0** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/pdi/job/PR-9078/5/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

File: plugins/put-a-file-with-ftp/impl/src/main/java/org/pentaho/di/job/entries/ftpput/JobEntryFTPPUT.java [hunk 2]
```diff
@@ -639,8 +642,16 @@ public Result execute( Result previousResult, int nr ) {
...
+      if( !Files.exists(path) ) {
+        throw new FTPException( BaseMessages.getString( PKG, "JobFTPPUT.FolderExists.NOK" ) );
+      }
+      if( !Files.isReadable( path ) || !Files.isWritable( path ) ) {
```
Comment (abryant-hv) [inline plugins/put-a-file-with-ftp/impl/src/main/java/org/pentaho/di/job/entries/ftpput/JobEntryFTPPUT.java:649]: inconsistent formatting. `if` is not a function.

**Comment by buildguy** [comment]
### :warning: Build finished in 7m 31s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl \
plugins/put-a-file-with-ftp/impl
```
---
### :no_entry: Failed Tests

#### :cloud_with_lightning_and_rain: 1 test(s) failed:

<details><summary><b>org.pentaho.di.job.entries.ftpput.JobEntryFTPPUTLoadSaveTest.org.pentaho.di.job.entries.ftpput.JobEntryFTPPUTLoadSaveTest</b> (click to expand)</summary><p>

```
${result.errorDetails}
```

</p></details>

  
Tests run: **7**, Failures: **1**, Skipped: **0** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/pdi/job/PR-9078/7/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 14m 37s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl \
plugins/put-a-file-with-ftp/impl
```
---
:ok_hand: All tests passed!
 
Tests run: **7**, Failures: **0**, Skipped: **0** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/pdi/job/PR-9078/8/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Review by abryant-hv — APPROVED** [review]

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 7m 32s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl \
plugins/put-a-file-with-ftp/impl
```
---
:ok_hand: All tests passed!
 
Tests run: **7**, Failures: **0**, Skipped: **0** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/pdi/job/PR-9078/9/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by NJtwentyone** [comment]
@Mandakini111  This can be approved and merged by another developer without addressing my previous comment (https://github.com/pentaho/pentaho-kettle/pull/9078#discussion_r1371935215)

I see that you have made some changes, since I last looked at it. I don't have the current bandwidth to look at it.

My attention is on other sprint obligations. I plan to look at it later today or this coming Monday.

File: plugins/put-a-file-with-ftp/impl/src/test/java/org/pentaho/di/job/entries/ftpput/JobEntryFTPPUTTest.java [hunk 3]
```diff
@@ -89,6 +114,43 @@ public void createFtpClient_UsesTimeoutIfSet() throws Exception {
...
+  public void initialSetUp() throws Exception {
+    JobEntryBase base = mock( JobEntryBase.class );
+    entry.setUserName( "username" );
+    entry.setPassword( "password" );
+    entry.setRemoteDirectory( "Remotedir" );
+    entry.setLocalDirectory( "Localdir" );
+    when( base.environmentSubstitute( "username" ) ).thenReturn( "mockedUsername" );
+    when( base.environmentSubstitute( "password" ) ).thenReturn( "mockedEncryptedPassword" );
+    try ( MockedStatic<Encr> encr = mockStatic( Encr.class ) ) {
+      encr.when( () -> Encr.decryptPasswordOptionallyEncrypted( anyString() ) )
```
Comment (NJtwentyone) [inline plugins/put-a-file-with-ftp/impl/src/test/java/org/pentaho/di/job/entries/ftpput/JobEntryFTPPUTTest.java:147]: This is throwing an NullPointerException when I attach a debugger to it.

Screenshot 2023-11-03 at 1 08 04 PM

File: plugins/put-a-file-with-ftp/impl/src/test/java/org/pentaho/di/job/entries/ftpput/JobEntryFTPPUTTest.java [hunk 4]
```diff
@@ -89,6 +114,43 @@ public void createFtpClient_UsesTimeoutIfSet() throws Exception {
...
 
+  @Test
+  public void testLocalDirDoesntExist() throws Exception{
+    this.initialSetUp();
+    Result result = entry.execute( previousResult, 1 );
+    assertEquals( 1, result.getNrErrors() );
+  }
+
+  @Test
+  public void testNoFilePermissions() throws Exception {
```
Comment (NJtwentyone) [inline plugins/put-a-file-with-ftp/impl/src/test/java/org/pentaho/di/job/entries/ftpput/JobEntryFTPPUTTest.java:125]: This test case `testNoFilePermissions()` doesn't make any sense.

 The new code you added throws an exception
`throw new FTPException(...`
But yet here you aren't checking for the thrown FTOException anywhere.

And somehow, the `execute()` return with a result.

File: plugins/put-a-file-with-ftp/impl/src/test/java/org/pentaho/di/job/entries/ftpput/JobEntryFTPPUTTest.java [hunk 5]
```diff
@@ -89,6 +114,43 @@ public void createFtpClient_UsesTimeoutIfSet() throws Exception {
     verify( ftpClient ).setTimeout( 10 );
   }
 
+  @Test
+  public void testLocalDirDoesntExist() throws Exception{
```
Comment (NJtwentyone) [inline plugins/put-a-file-with-ftp/impl/src/test/java/org/pentaho/di/job/entries/ftpput/JobEntryFTPPUTTest.java:118]: This test case `testLocalDirDoesntExist()` doesn't make any sense.

 The new code you added throws an exception
`throw new FTPException(...`
But yet here you aren't checking for the thrown FTOException anywhere.

And somehow, the `execute()` return with a result.

**Review by NJtwentyone — CHANGES_REQUESTED** [review]
Please remove your recent additions to `JobEntryFTPPUTTest.java`.

Thanks for taking the time to try to add some unit tests, but the additions aren't really testing the new logic properly. No fault to your own, some of the legacy Pentaho code doesn't allow for easy testing. On closer inspection, I have concerns the current unit tests aren't doing anything either, since they are using `Mockito.spy`. 

In order to get this class `JobEntryFTPPUT.java` unit testable, some refactoring will need to take place. This will need to be done by an experience developer. The end result will not use `Mockito.spy` for `JobEntryFTPPUT.java`

File: plugins/put-a-file-with-ftp/impl/src/main/java/org/pentaho/di/job/entries/ftpput/JobEntryFTPPUT.java [hunk 6]
```diff
@@ -639,8 +642,16 @@ public Result execute( Result previousResult, int nr ) {
         }
       }
 
+      Path path = Paths.get( realLocalDirectory );
+      if ( !Files.exists( path ) ) {
+        throw new FTPException( BaseMessages.getString( PKG, "JobFTPPUT.FolderExists.NOK" ) );
```
Comment (NJtwentyone) [inline plugins/put-a-file-with-ftp/impl/src/main/java/org/pentaho/di/job/entries/ftpput/JobEntryFTPPUT.java:647]: This code is still inside a big try/catch
```
try{
... ... ...
<< new throw exception code >>
... ... ...
} catch( Exception e){
      result.setNrErrors( 1 );
      logError( BaseMessages.getString( PKG, "JobFTPPUT.Log.ErrorPuttingFiles", e.getMessage() ) );
      logError( Const.getStackTracker( e ) );
} finally {
... ... ..
}
```

was the issue before the files weren't being copied over and failing silently and now thrown exception will get to the code line
`result.setNrErrors( 1 );`
which will satisfy the reporter's problem of "no errors will be reported and the job will be successfully completed." ?

File: plugins/put-a-file-with-ftp/impl/src/main/java/org/pentaho/di/job/entries/ftpput/JobEntryFTPPUT.java (hunk 1 above)
Comment (NJtwentyone) [reply plugins/put-a-file-with-ftp/impl/src/main/java/org/pentaho/di/job/entries/ftpput/JobEntryFTPPUT.java:645]: fixed with https://github.com/pentaho/pentaho-kettle/commit/42a31cf00e0008515d6a51939fcac7cd624e1075

File: plugins/put-a-file-with-ftp/impl/src/test/java/org/pentaho/di/job/entries/ftpput/JobEntryFTPPUTTest.java (hunk 4 above)
Comment (NJtwentyone) [reply plugins/put-a-file-with-ftp/impl/src/test/java/org/pentaho/di/job/entries/ftpput/JobEntryFTPPUTTest.java:118]: fixed with https://github.com/pentaho/pentaho-kettle/commit/42a31cf00e0008515d6a51939fcac7cd624e1075

File: plugins/put-a-file-with-ftp/impl/src/test/java/org/pentaho/di/job/entries/ftpput/JobEntryFTPPUTTest.java (hunk 5 above)
Comment (NJtwentyone) [reply plugins/put-a-file-with-ftp/impl/src/test/java/org/pentaho/di/job/entries/ftpput/JobEntryFTPPUTTest.java]: obsolete case in source code was removed with https://github.com/pentaho/pentaho-kettle/commit/42a31cf00e0008515d6a51939fcac7cd624e1075

File: plugins/put-a-file-with-ftp/impl/src/main/java/org/pentaho/di/job/entries/ftpput/JobEntryFTPPUT.java (hunk 6 above)
Comment (NJtwentyone) [reply plugins/put-a-file-with-ftp/impl/src/main/java/org/pentaho/di/job/entries/ftpput/JobEntryFTPPUT.java]: I think I answered my own question looking closely at the source code

```
      result.setNrErrors( 1 );
      logError( BaseMessages.getString( PKG, "JobFTPPUT.Log.ErrorPuttingFiles", e.getMessage() ) );
```

**Review by NJtwentyone — APPROVED** [review]
Thank you for adding the unit test.
